### PR TITLE
Add a new API method resize_disk

### PR DESCRIPTION
**Pull request by thoHeinze**
* this change allows to resize volumes using the AWS API
* the change is analogous to the change inside the BOSH Openstack CPI: https://github.com/cloudfoundry/bosh-openstack-cpi-release/commit/d081938852ecae502c0ba285020996a4a0c0d01d
* to use this feature the bosh director configuration enable_cpi_resize_disk needs to be enabled

### CODE REVIEW

**Comment by thoHeinze** [comment]
@mrosecrance  Any chance somebody could a review of this change?

**Comment by mrosecrance** [comment]
It's prioritized to be looked at sometime in the next few weeks.

**Comment by friegger** [comment]
This feature is a game changer for workloads which huge disks that require to evacuate data when draining. 
Though, as far as I remember the bosh-agent still has to learn to resize the partition after the disk has been resized in the IaaS. 

Context:
https://github.com/cloudfoundry/bosh-openstack-cpi-release/issues/161
https://github.com/cloudfoundry/bosh-agent/issues/221

There was a PR that was closed by the author, plus a WIP commit:
https://github.com/cloudfoundry/bosh-agent/pull/233
https://github.com/cloudfoundry/bosh-agent/commit/1c1a19a95885814a057b7d53ff8e65846cd27d18

**Comment by mrosecrance** [comment]
Thanks for the context, does it make sense to merge this without a bosh-agent change? I've only taken a quick look but it looks like the sfdisk partitions will be [resized](https://github.com/cloudfoundry/bosh-agent/blob/084f6147c315f20b2b6b941b368cd0d71fa6c774/platform/disk/sfdisk_partitioner.go#L200) but parted and [ephemeral will not](https://github.com/cloudfoundry/bosh-agent/blob/084f6147c315f20b2b6b941b368cd0d71fa6c774/platform/disk/ephemeral_device_partitioner.go#L42). Without a change in the agent, I would think this would be a backslide in the desired outcome.

**Comment by mrosecrance** [comment]
@thoHeinze is making a bosh-agent pr to handle partition resize something you're willing to do?

**Comment by ramonskie** [comment]
depends on https://github.com/cloudfoundry/bosh-agent/pull/244

File: src/bosh_aws_cpi/spec/unit/resize_disk_spec.rb [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+      allow(Bosh::Clouds::Config.logger).to receive(:info)
+
+      @cloud.resize_disk('disk-id', 2048)
+
+      expect(Bosh::Clouds::Config.logger).to have_received(:info).with('Skipping resize of disk disk-id because current value 2 GiB is equal new value 2 GiB')
+    end
+  end
+
+  context 'when trying to resize disk to a new size with an not even size in MiB' do
+    it 'does not call extend on disk and writes to the log' do
```
Comment (rkoster) [inline src/bosh_aws_cpi/spec/unit/resize_disk_spec.rb:44]: Why should it not call extend when resizing a disk with an not even size?
Also the test at line 50 actually tests that the method is called.

**Review by rkoster — CHANGES_REQUESTED** [review]
@thoHeinze Thanks for the Pull Request, there was just one small test that could use some clarification. For the rest it all makes sense.
cc @camilalonart

File: src/bosh_aws_cpi/spec/unit/resize_disk_spec.rb (hunk 1 above)
Comment (thoHeinze) [reply src/bosh_aws_cpi/spec/unit/resize_disk_spec.rb:44]: @rkoster  Sorry for the late reply. I just mimicked with my tests the behaviour of the openstack CLI: https://github.com/cloudfoundry/bosh-openstack-cpi-release/blob/master/src/bosh_openstack_cpi/spec/unit/resize_disk_spec.rb#L44
I'm ok with removing this, I thought this was the expected behaviour.

Comment (rkoster) [reply src/bosh_aws_cpi/spec/unit/resize_disk_spec.rb:44]: @beyhan Do you have any context on where that test in the openstack cpi came from? It seems like the test does not match its description. The test explicitly tests for extend to be called: https://github.com/cloudfoundry/bosh-openstack-cpi-release/blob/master/src/bosh_openstack_cpi/spec/unit/resize_disk_spec.rb#L54

**Review by bgandon — APPROVED** [review]

File: src/bosh_aws_cpi/spec/unit/resize_disk_spec.rb (hunk 1 above)
Comment (beyhan) [reply src/bosh_aws_cpi/spec/unit/resize_disk_spec.rb:44]: @rkoster The error in the description seems to be a copy&paste failure from the previous test (https://github.com/cloudfoundry/bosh-openstack-cpi-release/blob/a8fa225aa75db1df3c40d62c843b00b68ba385e4/src/bosh_openstack_cpi/spec/unit/resize_disk_spec.rb#L33). I corrected the test description and also extended it to fit to the reality. The PR is: https://github.com/cloudfoundry/bosh-openstack-cpi-release/pull/240

File: src/bosh_aws_cpi/lib/cloud/aws/cloud_core.rb [hunk 2]
```diff
@@ -160,6 +160,29 @@ def detach_disk(instance_id, disk_id)
...
+        @logger.info("Skipping resize of disk #{disk_id} because current value #{actual_size_gib} GiB" \
+                     " is equal new value #{new_size_gib} GiB")
+      elsif actual_size_gib > new_size_gib
+        cloud_error("Cannot resize volume to a smaller size from #{actual_size_gib} GiB to #{new_size_gib} GiB")
+      else
+        attachments = volume.attachments
+        unless attachments.empty?
+          cloud_error("Cannot resize volume '#{disk_id}' it still has #{attachments.size} attachment(s)")
+        end
+        @volume_manager.extend_ebs_volume(volume, new_size_gib)
```
[the remark this answers is not in the document]
Comment (beyhan) [reply src/bosh_aws_cpi/lib/cloud/aws/cloud_core.rb:180]: We definitely should document that the value is rounded up to the next higher number. I couldn't find any place where this is documented and needed to look into the implementation. This is not nice.

[the remark this answers is not in the document]
Comment (bgandon) [reply src/bosh_aws_cpi/lib/cloud/aws/cloud_core.rb:180]: Indeed, there is no documentation about the CPI API implementation. But the `docs/` folder here in this repo could be a good place where to start adding some stuff. No big deal anyway.

File: src/bosh_aws_cpi/spec/unit/resize_disk_spec.rb (hunk 1 above)
Comment (thoHeinze) [reply src/bosh_aws_cpi/spec/unit/resize_disk_spec.rb]: @rkoster  @beyhan  Thanks for clarifying. I now renamed the description similar to cloudfoundry/bosh-openstack-cpi-release#240

**Comment by rkoster** [comment]
@thoHeinze @bgandon @beyhan Thanks all for working on this!

**Pull request merged**